Fix various issues with scrolling

mrobinson opened the pull request:
This change fixes three issues with scrolling:

 - In order for scrolls from script to properly scroll nodes in WebRender that might require a layout, Servo should first flush layout and send the display list to WebRender.
 - `overflow: scroll` elements were not scrollable from script. This change also refactors the detection of a potentially scrollable body in quirks mode in order to make it match the specification more closely.
- Enable script scrolls in Layout 2020 by adding an implementation of the scroll area query, which is necessary to ensure that an element does support scrolling.

Each of these changes is a separate commit here with corresponding test expectation updates.


Fixes #29659.
Fixes servo/webrender#3078.

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #29659 and servo/webrender#3078.
- [x] There are tests for these changes


<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

bors-servo commented:
:hourglass: Trying commit 588fece6fc171a9b9b035e987f91f4a3fa5b9e7a with merge b61696907ed5ec8983d604e985e268cd3709e0a8...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4788924303/jobs/8518688069)

mrobinson committed:
Scroll from script should trigger a reflow

Scrolling from script should flow layout and send a display list to
WebRender. This allows all of the scroll nodes to exist in WebRender
before asking it to move the node.

See https://gist.github.com/paulirish/5d52fb081b3570c81e3a.
Fixes #29659.

mrobinson committed:
Allow script to scroll `overflow: scroll` elements

Before the code was only allowing `overflow: hidden` elements to scroll.
This fixes that issue and also clean up the code that deals with
detecting whether the body is a "potentially scrollable" in quirks mode.

mrobinson committed:
Fix scrolling from script in Layout 2020

Script will only scroll if it detects that an element has a scrolling
box, so this change adds an implementation of the scroll area query to
Layout 2020. This allows some scrolling tests to start passing.

This change also updates all expected results in css-backgrounds and
cssom-view.

mrobinson commented:
@bors-servo try

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4794845844/jobs/8530736630)

bors-servo commented:
:hourglass: Trying commit 9e6834d0ce743ea1e1701dc05531558f5823030c with merge df7ad5abdc30eee0b67b6dd53f923d310d396b3d...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/4796517939/jobs/8534418312)
State: approved= try=True

mrobinson commented:
@bors-servo r=mukilan

Thank you for the review!

bors-servo commented:
:pushpin: Commit 404ee8b has been approved by `mukilan`

<!-- @bors-servo r=mukilan 404ee8b98468970f0108d1e011f44b694975acb7 -->

bors-servo commented:
:hourglass: Testing commit 404ee8b98468970f0108d1e011f44b694975acb7 with merge e4d08358ac1a8749f6cf95bc017cbd87195d2981...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/4802100498/jobs/8547176958)
Approved by: mukilan
Pushing e4d08358ac1a8749f6cf95bc017cbd87195d2981 to master...
